hashing/src/lib.rs

```diff
@@ -111,14 +134,32 @@ impl Digest {
     /// l
     /// ```
     ///
+    /// Finally hashes the number of elements resulting hash. In the example above the final output
+    /// would be `hash_pair(u64_as_slice(6), l)`.
+    ///
     /// Returns [`Digest::SENTINEL_MERKLE_TREE`] when the input is empty.
     ///
     /// [1]: https://en.wikipedia.org/wiki/Merkle_tree
     /// [2]: https://en.wikipedia.org/wiki/Graph_reduction
-    pub fn hash_vec_merkle_tree(vec: Vec<Digest>) -> Digest {
-        vec.into_iter()
-            .tree_fold1(|x, y| Digest::hash_pair(&x, &y))
-            .unwrap_or(Self::SENTINEL_MERKLE_TREE)
+    pub fn hash_merkle_tree<I>(leaves: I) -> Digest
+    where
+        I: IntoIterator<Item = Digest>,
+    {
+        let (leaf_count, raw_root) = leaves
+            .into_iter()
+            .map(|x| (1u64, x))
```

Comment:
Can we get rid of these counters if we add a `?Sized` constraint? CC @afck

Reply:
Hmm maybe not if we want to do `hash_merkle_tree(data.chunks(1usize << 10).map(blake2b))`...

Reply:
You mean an `ExactSizeIterator` constraint? That should work, yes, even with your example.

Reply:
Implemented using `ExactSizeIterator`.